src/mail-app/mail/view/MailViewModel.ts

```diff
@@ -377,29 +381,43 @@ export class MailViewModel {
 		this._filterType = null
 
 		this._folder = folder
-		this.listStreamSubscription?.end(true)
-		this.listStreamSubscription = this.listModel!.stateStream.map((state) => this.onListStateChange(state))
-		this.listModel!.loadInitial().then(() => {
-			if (this.listModel != null && this._folder === folder) {
-				this.fixCounterIfNeeded(folder, this.listModel.items)
-			}
-		})
 	}
 
 	getConversationViewModel(): ConversationViewModel | null {
 		return this.conversationViewModel
 	}
 
-	private listModelForFolder = memoized((_folderId: Id) => {
+	private getListModel = memoized((_folderId: Id, groupMailsByConversation: boolean): MailSetListModel => {
 		// Capture state to avoid race conditions.
 		// We need to populate mail set entries cache when loading mails so that we can react to updates later.
 		const folder = assertNotNull(this._folder)
-		return new MailListModel(folder, this.conversationPrefProvider, this.entityClient, this.mailModel, this.inboxRuleHandler, this.cacheStorage)
+
+		let listModel: MailSetListModel
+		if (groupMailsByConversation && !MAIL_LIST_FOLDERS.includes(folder.folderType as MailSetKind)) {
+			listModel = new ConversationListModel(
+				folder,
+				this.conversationPrefProvider,
+				this.entityClient,
+				this.mailModel,
+				this.inboxRuleHandler,
+				this.cacheStorage,
+			)
+		} else {
+			listModel = new MailListModel(folder, this.conversationPrefProvider, this.entityClient, this.mailModel, this.inboxRuleHandler, this.cacheStorage)
+		}
+		this.listStreamSubscription?.end(true)
+		this.listStreamSubscription = listModel.stateStream.map((state: ListState<Mail>) => this.onListStateChange(listModel, state))
+		listModel!.loadInitial().then(() => {
```

Comment:
why do we need to force unwrap it? we also do not force unwrap in production code

Reply:
>[
we shouldn't need it now because it's passed in tho?